shesmu-server-ui/src/help.ts

```diff
@@ -0,0 +1,223 @@
+      hotspot: "stats-histogram",
+      area: "action",
+    },
+    {
+      summary: "Use the breakdown table to filter",
+      description:
+        "If there are many actions, a breakdown is provided with common alert attributes to allow fast filtering.",
+      hotspot: null,
+      area: "alert",
+    },
+    {
+      summary: "Custom actions can be uploaded",
+      description:
+        "The simulator has access to all the actions on a server, but the Extra Definitions tab allows uploading custom actions, for in-development workflows.",
+      hotspot: null,
+      area: "simulator",
+    },
+  ],
+];
+
+/**
+ * This is the help information stored in the browser to track what the user has seen.
+ */
+interface HelpState {
+  /** This is one above the highest version number the user has been exposted to. (This means it should be the length of the tips list once the system is initialised.)*/
+  version: number;
+  /**
+   * All the tips the user has yet to see. These are a pair of indices into the features list
+   */
+  tips: [number, number][];
+}
+let state = locallyStored<HelpState>("shesmu_help", { version: 0, tips: [] });
+let model: Publisher<HelpState> = pubSubModel();
+model.subscribe({ ...state.model, isAlive: true });
+{
+  const current = state.last;
+  // In case there are any garbage tips, remove them from the user's config. If the server is rolled back, and then forward, it might cause tips to be marked as read, but that's not such a big deal. It's also per server; not global accross all Shesmu instances a user accesses.
+  current.tips = current.tips.filter(
+    ([version, index]) =>
+      version >= 0 &&
+      version < tips.length &&
+      index >= 0 &&
+      index < tips[version].length
+  );
+  // Now, pack any unseen tips into the user's config
+  for (let i = current.version; i < tips.length; i++) {
+    current.tips = current.tips.concat(
+      [...tips[i].keys()].map((index) => [i, index])
+    );
+  }
+  current.version = tips.length;
+  model.statusChanged(current);
+}
+
+function helpButton(
+  predicate: (version: number, index: number, tip: HelpTip) => boolean
+): UIElement {
+  return subscribedState(state.last, model, (input: HelpState) => {
+    const hasNew = input.tips.some(([version, index]) =>
+      predicate(version, index, tips[version]?.[index])
+    );
+    const popupContainer = document.createElement("div");
+    popupContainer.className = "helpcapture";
+    const popup = document.createElement("div");
+    addElements(
+      popup,
+      buttonIcon("✓ Read", "Mark all these tips as read.", () => {
+        document.removeChild(popupContainer);
+        model.statusChanged({
+          version: input.version,
+          tips: input.tips.filter(
+            ([version, index]) =>
+              !predicate(version, index, tips[version]?.[index])
+          ),
+        });
+      }),
+      " | ",
+      buttonIcon("💯 Read All", "Mark all tips as read.", () => {
```

Comment:
I'm confused about the distinction between "mark all _these_ tips as read" vs "mark all tips as read"

Reply:
So, in the usual tip systems, what I really want is "go die in a fire" button and I'm trying to phrase that delicately.

Reply:
![image](https://user-images.githubusercontent.com/17747491/86785184-57aadc80-c030-11ea-8ba0-5a899e476c68.png)
So on this page, there is only 1 tip, so I'd expect 'Read' and 'Read All' to do the same action. 'Read All' would actually mark off _all_ tips, in all the boxes across Shesmu, until the next version is released? In that case I'd recommend renaming Read All to something like 'Turn off help notifications' and maybe move it to the navbar or somewhere else

Reply:
Okay, I'll make it visible on the main Shesmu status page only. I've upgraded the tooltips on read to provide some more context too.

Reply:
Opening the tooltip on the main page then clicking away (without clicking Read or Read All) clears the notifications for all of shesmu. Also, the naming is still confusing because 'Read All' being _inside_ the tooltip implies those actions only apply within the context of that tooltip.

Reply:
Yeah, that was a mis-feature by design. I've updated the code to track which have been read and then only clear those.